### PR TITLE
feat: support custom VM domain

### PR DESCRIPTION
**Pull request by moul**
Introducing the concept of "ChainDomain," a local primary domain for packages.

The next step, which will be another PR, is to ensure that we can launch a gnoland/gnodev instance while importing a local folder or using a genesis to preload packages from other domains. This will allow users to add packages only to the primary domain while accessing packages from multiple domains. The result will be a preview of the upcoming IBC era, where a single chain can add packages only to its domain but can fetch missing dependencies from other registered zones.

- [x] gnovm unaware of gno.land, just accepting valid domains
- [x] vmkeeper initialized with a domain
- [x] Stdlib to know the current primary domain + new std.ChainDomain
- [x] new unit tests around custom domains

Depends on #2910 
Depends on https://github.com/gnolang/gno/pull/3003
Blocks a new PR that will add multidomain support.
Related with https://github.com/gnolang/gno/issues/2904#issuecomment-2395011435

### CODE REVIEW

File: gno.land/cmd/gnoland/start.go
Comment (moul) [inline gno.land/cmd/gnoland/start.go:140]: (deadcode)

File: gno.land/pkg/gnoland/app.go
Comment (n0izn0iz) [inline gno.land/pkg/gnoland/app.go:228]: why add this here instead of in genesis?

File: gno.land/pkg/sdk/vm/keeper.go
Comment (n0izn0iz) [inline gno.land/pkg/sdk/vm/keeper.go:93]:
```suggestion
	vmk.reNamespace = regexp.MustCompile(`^` + regexp.QuoteMeta(chainDomain) + `/(?:r|p)/([\.~_a-zA-Z0-9]+)`)
```
otherwise it will also match `gnoxland` for `chainDomain == "gno.land"`

**Review by n0izn0iz — CHANGES_REQUESTED** [review]

File: gno.land/pkg/gnoland/app.go
Comment (moul) [reply gno.land/pkg/gnoland/app.go:228]: I think that I prefer your idea. I'm waiting for more feedback before making the change.

Comment (thehowl) [reply gno.land/pkg/gnoland/app.go:227]: Yeah I agree it should be in genesis. Allows us more easily to find problems when there's a mismatch between two nodes.

File: gnovm/pkg/gnolang/helpers.go
Comment (thehowl) [inline gnovm/pkg/gnolang/helpers.go:15]: Should use this to validate the chain domain elsewhere.

File: gnovm/pkg/gnolang/helpers.go
Comment (thehowl) [inline gnovm/pkg/gnolang/helpers.go:36]:
```
return strings.IndexByte(s[:strings.IndexByte(s, '/')+1], '.') < 0
```

File: gnovm/stdlibs/std/native.go
Comment (thehowl) [inline gnovm/stdlibs/std/native.go:31]: Since this is used to write the import paths, should this not be largely known by gno files already (so no need for the function)?

**Review by thehowl — COMMENTED** [review]
Maybe look at changing this for `gno.land/pkg/gnoweb/gnoweb.go` as well. No other significant places breaking that I can see, at least in our codebase.

**Comment by thehowl** [comment]
I like the feature. I shared it with the core contribs so we have some warning ahead of time of what we want to do :)

File: gnovm/stdlibs/std/native.go
Comment (moul) [reply gnovm/stdlibs/std/native.go:31]: It's for when your `p/` package is imported from one chain to another. The code remains the same, but the behavior may change.

File: gno.land/pkg/gnoland/app.go
Comment (moul) [reply gno.land/pkg/gnoland/app.go:227]: switching the PR to draft until I finish and merge #3003.

**Comment by moul** [comment]
Draft because depending on https://github.com/gnolang/gno/pull/3003

File: gnovm/stdlibs/std/native.gno
Comment (albttx) [inline gnovm/stdlibs/std/native.gno:14]: There was a discussion on `PrevRealm` that was about removing all the `Get` prefixes, is it's still accurate ?

In this case, `std.ChainID`, `std.ChainDomain` and `std.Height`

Comment (moul) [reply gnovm/stdlibs/std/native.gno:14]: Yep, we'll remove in a refactor, right now it makes sense to just be consistent.

File: gnovm/pkg/gnolang/helpers.go
Comment (moul) [reply gnovm/pkg/gnolang/helpers.go]: c9f229270

File: gno.land/pkg/gnoland/app.go
Comment (moul) [reply gno.land/pkg/gnoland/app.go]: done.

File: gno.land/pkg/gnoland/node_inmemory.go
Comment (MikaelVallenet) [inline gno.land/pkg/gnoland/node_inmemory.go:42]:
```suggestion
	_ = domainParam.Parse("gno.land/r/sys/params.vm.chain.domain.string=" + chaindomain)
```

File: gno.land/cmd/gnoland/testdata/genesis_params.txtar
Comment (MikaelVallenet) [inline gno.land/cmd/gnoland/testdata/genesis_params.txtar:10]:
```suggestion
gnokey query params/vm/gno.land/r/sys/params.vm.chain.domain.string
```

File: gno.land/genesis/genesis_params.toml
Comment (MikaelVallenet) [inline gno.land/genesis/genesis_params.toml:11]:
```suggestion
  chain.domain.string = "gno.land"
```

Comment (MikaelVallenet) [reply gno.land/genesis/genesis_params.toml:11]: could be applied to commented params below

**Review by MikaelVallenet — COMMENTED** [review]
i also found that it does not work with gnodev for now, the params should be added to genesis state in ``contribs/gnodev/pkg/dev/node.go`` and added into every execContext in ``gno.land/pkg/sdk/vm/keeper.go`` (it's what i did on my side with chain tz to make it work with gnodev, maybe it's not the good way)

File: gno.land/genesis/genesis_params.toml
Comment (moul) [reply gno.land/genesis/genesis_params.toml:11]: Oh, sorry. The category is "vm," not "chain." Let's use an underscore everywhere. Thanks.

File: gno.land/pkg/sdk/vm/params.go
Comment (moul) [reply gno.land/pkg/sdk/vm/params.go:7]: sorry, was wrong, should be an underscore.

**Comment by moul** [comment]
> i also found that it does not work with gnodev for now, the params should be added to genesis state in contribs/gnodev/pkg/dev/node.go and added into every execContext in gno.land/pkg/sdk/vm/keeper.go (it's what i did on my side with chain tz to make it work with gnodev, maybe it's not the good way)

Do you have an example?

**Comment by MikaelVallenet** [comment]
> > i also found that it does not work with gnodev for now, the params should be added to genesis state in contribs/gnodev/pkg/dev/node.go and added into every execContext in gno.land/pkg/sdk/vm/keeper.go (it's what i did on my side with chain tz to make it work with gnodev, maybe it's not the good way)
> 
> Do you have an example?

So here is the addition of params into genesis: https://github.com/gnolang/gno/pull/3193/commits/c3ecc9706d7a0f0db509af2d16639a7cd2f77f1c
Here is the file where i add chain tz to every ExecContext: https://github.com/gnolang/gno/pull/3193/files#diff-3663e30f0c0e01d8d47bb7974d605b4cf12d1a244970c6a9bbba55fb0be6aced
Without this, running ``gnodev --<param>=<value>`` will not work, (not sure it's good to do the way i did tbh)

**Comment by moul** [comment]
> So here is the addition of params into genesis: https://github.com/gnolang/gno/commit/c3ecc9706d7a0f0db509af2d16639a7cd2f77f1c

I suggest you switch to my method. I implemented this in `gnoland`, which is used by `pkg/integration`. You made a similar change, but only in `gnodev`, making this feature harder to use and test outside of `gnodev`.

> Here is the file where i add chain tz to every ExecContext: [#3193 (files)](https://github.com/gnolang/gno/pull/3193/files#diff-3663e30f0c0e01d8d47bb7974d605b4cf12d1a244970c6a9bbba55fb0be6aced)

If I'm not mistaken, that's what I did as well. Do you see any places where I overlooked it?

**Comment by MikaelVallenet** [comment]
> > So here is the addition of params into genesis: [c3ecc97](https://github.com/gnolang/gno/commit/c3ecc9706d7a0f0db509af2d16639a7cd2f77f1c)
> 
> I suggest you switch to my method. I implemented this in `gnoland`, which is used by `pkg/integration`. You made a similar change, but only in `gnodev`, making this feature harder to use and test outside of `gnodev`.
> 
> > Here is the file where i add chain tz to every ExecContext: [#3193 (files)](https://github.com/gnolang/gno/pull/3193/files#diff-3663e30f0c0e01d8d47bb7974d605b4cf12d1a244970c6a9bbba55fb0be6aced)
> 
> If I'm not mistaken, that's what I did as well. Do you see any places where I overlooked it?

- Yes this one is just the commit about gnodev, but i think the change should be applied to both places

- I can see on your PR 
you did not add ChainDomain in ExecContext return value in  ``QueryEval`` and ``QueryEvalString`` and ``Call`` function maybe ``AddPackage`` function too (github diff is a bit broken and does not display func name). I added it to these places, but i dont know exactly what QueryEval or QueryEval are used for, just felt it could fix my gnodev issue since it returns ExecContext and it works

**Comment by moul** [comment]
> Yes this one is just the commit about gnodev, but i think the change should be applied to both places

```
`gnodev` -> `gno.land/pkg/integration` > `gno.land/pkg/gnold`
    CLI       ->             helper with new arg  >        actual place where we create a dynamic genesis
```

The `param` override should occur in `gnoland`. `gnodev` should simply configure its CLI parser and pass the new option, but it shouldn't manipulate the genesis directly (not for this case).

> I can see on your PR
you did not add ChainDomain in ExecContext return value in QueryEval and QueryEvalString and Call function maybe AddPackage function too (github diff is a bit broken and does not display func name). I added it to these places, but i dont know exactly what QueryEval or QueryEval are used for, just felt it could fix my gnodev issue since it returns ExecContext and it works

I should definitely add them. Thank you.

**Review by MikaelVallenet — APPROVED** [review]

File: gnovm/pkg/gnolang/helpers.go
Comment (thehowl) [inline gnovm/pkg/gnolang/helpers.go:40]: `"morgan.land"` would be a standard library, best to consider whole of `s` if IndexByte < 0

File: gno.land/pkg/sdk/vm/keeper.go
Comment (thehowl) [inline gno.land/pkg/sdk/vm/keeper.go:196]: probably none at all

**Review by thehowl — APPROVED** [review]
yeah!

File: gnovm/pkg/gnolang/helpers.go
Comment (moul) [reply gnovm/pkg/gnolang/helpers.go:40]:
```suggestion
	idx := strings.IndexByte(s, '/')
	if idx < 0 {
		// If no '/' is found, consider the whole string
		return strings.IndexByte(s, '.') < 0
	}
	return strings.IndexByte(s[:idx+1], '.') < 0
```

**Pull request merged**
